File: jeantrail_project/src/services/plugins.ts
// Plugin System Service Interface and Implementation
export interface Plugin {
  id: string;
  name: string;
  displayName: string;
  description?: string;
  version: string;
  author?: string;
  manifest: PluginManifest;
  entryPoint?: string;
  permissions: string[];
  isActive: boolean;
  isSystem: boolean;
  installedAt: Date;
  updatedAt: Date;
  createdBy?: string;
}

export interface PluginManifest {
  name: string;
  version: string;
  description?: string;
  author?: string;
  homepage?: string;
  repository?: string;
  license?: string;
  keywords: string[];
  entry: string;
  permissions: PluginPermission[];
  apiVersion: string;
  minJeantrailVersion: string;
  maxJeantrailVersion?: string;
  dependencies: Record<string, string>;
  resources: PluginResources;
}

export interface PluginPermission {
  name: string;
  description: string;
  required: boolean;
}

export interface PluginResources {
  cssFiles: string[];
  jsFiles: string[];
  assets: string[];
  icons: PluginIcons;
}

export interface PluginIcons {
  icon16x16?: string;
  icon32x32?: string;
  icon48x48?: string;
  icon128x128?: string;
}

export interface UserPluginSettings {
  id: string;
  userId: string;
  pluginId: string;
  settings: Record<string, any>;
  isEnabled: boolean;
  lastUsedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

export interface CreatePluginRequest {
  name: string;
  displayName: string;
  description?: string;
  version: string;
  author?: string;
  manifest: PluginManifest;
  entryPoint?: string;
  permissions: string[];
}

export interface UpdatePluginRequest {
  displayName?: string;
  description?: string;
  version?: string;
  manifest?: PluginManifest;
  entryPoint?: string;
  permissions?: string[];
  isActive?: boolean;
}

export interface UpdateUserPluginSettingsRequest {
  settings: Record<string, any>;
  isEnabled: boolean;
}

export interface InstallPluginRequest {
  url: string;
  autoActivate: boolean;
}

export interface PluginCommand {
  action: string;
  parameters: Record<string, any>;
}

export interface ListPluginsQuery {
  isActive?: boolean;
  isSystem?: boolean;
  author?: string;
}

export interface PluginService {
  // Plugin management
  getPlugins: (query?: ListPluginsQuery) => Promise<Plugin[]>;
  getPlugin: (id: string) => Promise<Plugin>;
  createPlugin: (request: CreatePluginRequest) => Promise<Plugin>;
  updatePlugin: (id: string, request: UpdatePluginRequest) => Promise<Plugin>;
  deletePlugin: (id: string) => Promise<void>;
  activatePlugin: (id: string) => Promise<{ success: boolean; message: string }>;
  deactivatePlugin: (id: string) => Promise<{ success: boolean; message: string }>;
  
  // Plugin installation
  installPluginFromUrl: (request: InstallPluginRequest) => Promise<{ success: boolean; pluginId: string }>;
  validatePluginManifest: (manifest: PluginManifest) => Promise<{ valid: boolean; errors: string[] }>;
  
  // User plugin settings
  getUserPluginSettings: (userId: string, pluginId: string) => Promise<UserPluginSettings>;
  updateUserPluginSettings: (userId: string, pluginId: string, request: UpdateUserPluginSettingsRequest) => Promise<UserPluginSettings>;
  
  // Plugin execution
  executePluginCommand: (pluginId: string, userId: string, command: PluginCommand) => Promise<{
    success: boolean;
    result: string;
    commandId: string;
  }>;
  
  // Plugin API bridge
  bridgePluginApiRequest: (pluginId: string, userId: string, apiEndpoint: string, request: Record<string, any>) => Promise<{
    success: boolean;
    data: string;
  }>;
}

class PluginServiceImpl implements PluginService {
  private baseUrl: string;

  constructor(baseUrl: string = 'http://localhost:8080') {
    this.baseUrl = baseUrl;
  }

  async getPlugins(query?: ListPluginsQuery): Promise<Plugin[]> {
    const params = new URLSearchParams();
    if (query?.isActive !== undefined) params.append('isActive', query.isActive.toString());
    if (query?.isSystem !== undefined) params.append('isSystem', query.isSystem.toString());
    if (query?.author) params.append('author', query.author);

    const response = await fetch(`${this.baseUrl}/api/plugins?${params}`);
    if (!response.ok) throw new Error(`Failed to fetch plugins: ${response.statusText}`);
    return response.json();
  }

  async getPlugin(id: string): Promise<Plugin> {
    const response = await fetch(`${this.baseUrl}/api/plugins/${id}`);
    if (!response.ok) throw new Error(`Failed to fetch plugin: ${response.statusText}`);
    return response.json();
  }

  async createPlugin(request: CreatePluginRequest): Promise<Plugin> {
    const response = await fetch(`${this.baseUrl}/api/plugins`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to create plugin: ${response.statusText}`);
    return response.json();
  }

  async updatePlugin(id: string, request: UpdatePluginRequest): Promise<Plugin> {
    const response = await fetch(`${this.baseUrl}/api/plugins/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to update plugin: ${response.statusText}`);
    return response.json();
  }

  async deletePlugin(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/api/plugins/${id}`, {
      method: 'DELETE',
    });
    if (!response.ok) throw new Error(`Failed to delete plugin: ${response.statusText}`);
  }

  async activatePlugin(id: string): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${this.baseUrl}/api/plugins/${id}/activate`, {
      method: 'POST',
    });
    if (!response.ok) throw new Error(`Failed to activate plugin: ${response.statusText}`);
    return response.json();
  }

  async deactivatePlugin(id: string): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${this.baseUrl}/api/plugins/${id}/deactivate`, {
      method: 'POST',
    });
    if (!response.ok) throw new Error(`Failed to deactivate plugin: ${response.statusText}`);
    return response.json();
  }

  async installPluginFromUrl(request: InstallPluginRequest): Promise<{ success: boolean; pluginId: string }> {
    const response = await fetch(`${this.baseUrl}/api/plugins/install`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to install plugin: ${response.statusText}`);
    return response.json();
  }

  async validatePluginManifest(manifest: PluginManifest): Promise<{ valid: boolean; errors: string[] }> {
    const response = await fetch(`${this.baseUrl}/api/plugins/validate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(manifest),
    });
    if (!response.ok) throw new Error(`Failed to validate plugin manifest: ${response.statusText}`);
    return response.json();
  }

  async getUserPluginSettings(userId: string, pluginId: string): Promise<UserPluginSettings> {
    const response = await fetch(`${this.baseUrl}/api/plugins/users/${userId}/${pluginId}`);
    if (!response.ok) throw new Error(`Failed to fetch user plugin settings: ${response.statusText}`);
    return response.json();
  }

  async updateUserPluginSettings(userId: string, pluginId: string, request: UpdateUserPluginSettingsRequest): Promise<UserPluginSettings> {
    const response = await fetch(`${this.baseUrl}/api/plugins/users/${userId}/${pluginId}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to update user plugin settings: ${response.statusText}`);
    return response.json();
  }

  async executePluginCommand(pluginId: string, userId: string, command: PluginCommand): Promise<{
    success: boolean;
    result: string;
    commandId: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/plugins/${pluginId}/users/${userId}/execute`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(command),
    });
    if (!response.ok) throw new Error(`Failed to execute plugin command: ${response.statusText}`);
    return response.json();
  }

  async bridgePluginApiRequest(pluginId: string, userId: string, apiEndpoint: string, request: Record<string, any>): Promise<{
    success: boolean;
    data: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/plugins/${pluginId}/users/${userId}/api/${apiEndpoint}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to bridge plugin API request: ${response.statusText}`);
    return response.json();
  }
}

// Singleton instance
export const pluginService = new PluginServiceImpl();

// React hook
export const usePluginService = (): PluginService => {
  return pluginService;
};

// Plugin sandbox implementation
export class PluginSandbox {
  private pluginId: string;
  private userId: string;
  private allowedPermissions: string[];

  constructor(pluginId: string, userId: string, allowedPermissions: string[]) {
    this.pluginId = pluginId;
    this.userId = userId;
    this.allowedPermissions = allowedPermissions;
  }

  // Check if plugin has permission for specific action
  hasPermission(permission: string): boolean {
    return this.allowedPermissions.includes(permission);
  }

  // Safe API call wrapper
  async safeApiCall(apiEndpoint: string, data: Record<string, any>): Promise<any> {
    if (!this.hasPermission(`api.${apiEndpoint}`)) {
      throw new Error(`Plugin does not have permission for API endpoint: ${apiEndpoint}`);
    }

    return pluginService.bridgePluginApiRequest(this.pluginId, this.userId, apiEndpoint, data);
  }

  // Generate sandbox iframe URL
  getSandboxUrl(entryPoint: string): string {
    const baseUrl = window.location.origin;
    return `${baseUrl}/plugins/${this.pluginId}${entryPoint}?userId=${this.userId}&sandbox=true`;
  }
}

// Helper functions
export function createPluginSandbox(plugin: Plugin, userId: string): PluginSandbox {
  return new PluginSandbox(plugin.id, userId, plugin.permissions);
}

export function validatePluginPermission(permission: string): boolean {
  const validPermissions = [
    'jean.chat',
    'jean.actions',
    'tabs.read',
    'tabs.write',
    'proxy.read',
    'proxy.write',
    'files.read',
    'files.write',
    'network.request',
    'storage.read',
    'storage.write',
    'ui.panel',
    'ui.modal',
    'notifications.show',
  ];

  return validPermissions.includes(permission);
}

export function getPluginPermissionDescription(permission: string): string {
  const descriptions: Record<string, string> = {
    'jean.chat': 'Access Jean AI chat functionality',
    'jean.actions': 'Execute Jean AI actions',
    'tabs.read': 'Read browser tab information',
    'tabs.write': 'Modify browser tabs',
    'proxy.read': 'Read proxy network status',
    'proxy.write': 'Control proxy network settings',
    'files.read': 'Read local files',
    'files.write': 'Write local files',
    'network.request': 'Make network requests',
    'storage.read': 'Read local storage data',
    'storage.write': 'Write local storage data',
    'ui.panel': 'Create UI panels',
    'ui.modal': 'Show modal dialogs',
    'notifications.show': 'Show notifications',
  };

  return descriptions[permission] || permission;
}

// Plugin development helpers
export function createPluginManifest(manifest: Partial<PluginManifest>): PluginManifest {
  return {
    name: manifest.name || '',
    version: manifest.version || '1.0.0',
    description: manifest.description,
    author: manifest.author,
    homepage: manifest.homepage,
    repository: manifest.repository,
    license: manifest.license,
    keywords: manifest.keywords || [],
    entry: manifest.entry || 'index.html',
    permissions: manifest.permissions || [],
    apiVersion: manifest.apiVersion || 'v1',
    minJeantrailVersion: manifest.minJeantrailVersion || '1.0.0',
    maxJeantrailVersion: manifest.maxJeantrailVersion,
    dependencies: manifest.dependencies || {},
    resources: manifest.resources || {
      cssFiles: [],
      jsFiles: [],
      assets: [],
      icons: {},
    },
  };
}

export function generatePluginTemplate(name: string): CreatePluginRequest {
  const manifest = createPluginManifest({
    name: name.toLowerCase().replace(/\s+/g, '-'),
    displayName: name,
    description: `A ${name} plugin for JeanTrail`,
    version: '1.0.0',
    author: 'JeanTrail Plugin Developer',
    entry: '/index.html',
    permissions: [],
  });

  return {
    name: manifest.name,
    displayName: name,
    description: manifest.description,
    version: manifest.version,
    author: manifest.author,
    manifest,
    entryPoint: manifest.entry,
    permissions: [],
  };
}

File: src-tauri/src/plugins.rs
// Plugin System Module
use axum::{Json, extract::{Path, Query, State}};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;
use chrono::{DateTime, Utc};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct Plugin {
    pub id: Uuid,
    pub name: String,
    pub display_name: String,
    pub description: Option<String>,
    pub version: String,
    pub author: Option<String>,
    pub manifest: Value,
    pub entry_point: Option<String>,
    pub permissions: Vec<String>,
    pub is_active: bool,
    pub is_system: bool,
    pub installed_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Option<Uuid>,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct UserPluginSettings {
    pub id: Uuid,
    pub user_id: Uuid,
    pub plugin_id: Uuid,
    pub settings: Value,
    pub is_enabled: bool,
    pub last_used_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
pub struct CreatePluginRequest {
    pub name: String,
    pub display_name: String,
    pub description: Option<String>,
    pub version: String,
    pub author: Option<String>,
    pub manifest: Value,
    pub entry_point: Option<String>,
    pub permissions: Vec<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdatePluginRequest {
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub version: Option<String>,
    pub manifest: Option<Value>,
    pub entry_point: Option<String>,
    pub permissions: Option<Vec<String>>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateUserPluginSettingsRequest {
    pub settings: Value,
    pub is_enabled: bool,
}

// Plugin Manifest Structure
#[derive(Debug, Serialize, Deserialize)]
pub struct PluginManifest {
    pub name: String,
    pub version: String,
    pub description: Option<String>,
    pub author: Option<String>,
    pub homepage: Option<String>,
    pub repository: Option<String>,
    pub license: Option<String>,
    pub keywords: Vec<String>,
    pub entry: String,
    pub permissions: Vec<PluginPermission>,
    pub api_version: String,
    pub min_jeantrail_version: String,
    pub max_jeantrail_version: Option<String>,
    pub dependencies: HashMap<String, String>,
    pub resources: PluginResources,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PluginPermission {
    pub name: String,
    pub description: String,
    pub required: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PluginResources {
    pub css_files: Vec<String>,
    pub js_files: Vec<String>,
    pub assets: Vec<String>,
    pub icons: PluginIcons,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PluginIcons {
    pub icon_16x16: Option<String>,
    pub icon_32x32: Option<String>,
    pub icon_48x48: Option<String>,
    pub icon_128x128: Option<String>,
}

pub async fn list_plugins(
    Query(params): Query<ListPluginsQuery>,
) -> Result<Json<Vec<Plugin>>, axum::http::StatusCode> {
    let mock_plugins = vec![
        Plugin {
            id: Uuid::new_v4(),
            name: "hello-ai".to_string(),
            display_name: "Hello AI Assistant".to_string(),
            description: Some("A simple AI assistant plugin that uses Jean API".to_string()),
            version: "1.0.0".to_string(),
            author: Some("JeanTrail Team".to_string()),
            manifest: serde_json::json!({
                "name": "hello-ai",
                "version": "1.0.0",
                "description": "Simple AI assistant plugin",
                "entry": "/plugins/hello-ai/index.html",
                "permissions": [
                    {
                        "name": "jean.chat",
                        "description": "Access Jean AI chat functionality",
                        "required": true
                    }
                ],
                "api_version": "v1",
                "min_jeantrail_version": "1.0.0"
            }),
            entry_point: Some("/plugins/hello-ai/index.html".to_string()),
            permissions: vec!["jean.chat".to_string()],
            is_active: true,
            is_system: false,
            installed_at: Utc::now(),
            updated_at: Utc::now(),
            created_by: Some(Uuid::new_v4()),
        }
    ];

    Ok(Json(mock_plugins))
}

#[derive(Debug, Deserialize)]
pub struct ListPluginsQuery {
    pub is_active: Option<bool>,
    pub is_system: Option<bool>,
    pub author: Option<String>,
}

pub async fn get_plugin(
    Path(id): Path<Uuid>,
) -> Result<Json<Plugin>, axum::http::StatusCode> {
    // Fetch specific plugin
    Err(axum::http::StatusCode::NOT_FOUND)
}

pub async fn create_plugin(
    Json(request): Json<CreatePluginRequest>,
) -> Result<Json<Plugin>, axum::http::StatusCode> {
    let new_plugin = Plugin {
        id: Uuid::new_v4(),
        name: request.name.clone(),
        display_name: request.display_name,
        description: request.description,
        version: request.version,
        author: request.author,
        manifest: request.manifest,
        entry_point: request.entry_point,
        permissions: request.permissions,
        is_active: false,
        is_system: false,
        installed_at: Utc::now(),
        updated_at: Utc::now(),
        created_by: Some(Uuid::new_v4()),
    };

    // Validate manifest and save to database
    Ok(Json(new_plugin))
}

pub async fn update_plugin(
    Path(id): Path<Uuid>,
    Json(request): Json<UpdatePluginRequest>,
) -> Result<Json<Plugin>, axum::http::StatusCode> {
    // Update plugin in database
    Err(axum::http::StatusCode::NOT_FOUND)
}

pub async fn delete_plugin(
    Path(id): Path<Uuid>,
) -> Result<(), axum::http::StatusCode> {
    // Delete plugin from database
    Ok(())
}

pub async fn activate_plugin(
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Activate plugin
    Ok(Json(serde_json::json!({
        "success": true,
        "message": "Plugin activated successfully"
    })))
}

pub async fn deactivate_plugin(
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Deactivate plugin
    Ok(Json(serde_json::json!({
        "success": true,
        "message": "Plugin deactivated successfully"
    })))
}

pub async fn get_user_plugin_settings(
    Path(user_id): Path<Uuid>,
    Path(plugin_id): Path<Uuid>,
) -> Result<Json<UserPluginSettings>, axum::http::StatusCode> {
    // Get user's plugin settings
    Err(axum::http::StatusCode::NOT_FOUND)
}

pub async fn update_user_plugin_settings(
    Path(user_id): Path<Uuid>,
    Path(plugin_id): Path<Uuid>,
    Json(request): Json<UpdateUserPluginSettingsRequest>,
) -> Result<Json<UserPluginSettings>, axum::http::StatusCode> {
    let settings = UserPluginSettings {
        id: Uuid::new_v4(),
        user_id,
        plugin_id,
        settings: request.settings,
        is_enabled: request.is_enabled,
        last_used_at: Some(Utc::now()),
        created_at: Utc::now(),
        updated_at: Utc::now(),
    };

    // Save to database
    Ok(Json(settings))
}

pub async fn install_plugin_from_url(
    Json(request): Json<InstallPluginRequest>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Download and install plugin from URL
    // Validate manifest
    // Extract files
    // Save to database
    
    Ok(Json(serde_json::json!({
        "success": true,
        "message": "Plugin installed successfully",
        "plugin_id": Uuid::new_v4()
    })))
}

#[derive(Debug, Deserialize)]
pub struct InstallPluginRequest {
    pub url: String,
    pub auto_activate: bool,
}

pub async fn validate_plugin_manifest(
    Json(manifest): Json<PluginManifest>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    let mut validation_errors = Vec::new();
    
    // Validate required fields
    if manifest.name.is_empty() {
        validation_errors.push("Plugin name is required");
    }
    
    if manifest.entry.is_empty() {
        validation_errors.push("Plugin entry point is required");
    }
    
    // Validate version format
    if !is_valid_version(&manifest.version) {
        validation_errors.push("Invalid version format");
    }
    
    // Validate permissions
    for permission in &manifest.permissions {
        if !is_valid_permission(&permission.name) {
            validation_errors.push(&format!("Invalid permission: {}", permission.name));
        }
    }
    
    Ok(Json(serde_json::json!({
        "valid": validation_errors.is_empty(),
        "errors": validation_errors
    })))
}

fn is_valid_version(version: &str) -> bool {
    // Simple semantic version validation
    version.split('.').count() == 3 && version.split('.').all(|part| part.parse::<u32>().is_ok())
}

fn is_valid_permission(permission: &str) -> bool {
    // List of valid permissions
    let valid_permissions = vec![
        "jean.chat",
        "jean.actions",
        "tabs.read",
        "tabs.write",
        "proxy.read",
        "proxy.write",
        "files.read",
        "files.write",
        "network.request",
        "storage.read",
        "storage.write",
        "ui.panel",
        "ui.modal",
        "notifications.show",
    ];
    
    valid_permissions.contains(&permission)
}

// Plugin Sandbox Execution
pub async fn execute_plugin_command(
    Path(plugin_id): Path<Uuid>,
    Path(user_id): Path<Uuid>,
    Json(command): Json<PluginCommand>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Validate user has plugin enabled
    // Check permissions
    // Execute command in sandbox
    
    Ok(Json(serde_json::json!({
        "success": true,
        "result": "Command executed successfully",
        "command_id": Uuid::new_v4()
    })))
}

#[derive(Debug, Deserialize)]
pub struct PluginCommand {
    pub action: String,
    pub parameters: Value,
}

// Plugin API Bridge
pub async fn bridge_plugin_api_request(
    Path(plugin_id): Path<Uuid>,
    Path(user_id): Path<Uuid>,
    Path(api_endpoint): Path<String>,
    Json(request): Json<Value>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Route plugin API requests to appropriate handlers
    // Apply permission checks
    // Log API calls for audit
    
    Ok(Json(serde_json::json!({
        "success": true,
        "data": "API response"
    })))
}